src/main/java/com/google/cloud/genomics/dataflow/readers/bam/Sharder.java

```diff
@@ -0,0 +1,204 @@
+    reader = r.reader;
+    indexStream = r.index;
+    header = reader.getFileHeader();
+    hasIndex = reader.hasIndex() && reader.indexing().hasBrowseableIndex();
+    LOG.info("Has index = " + hasIndex);
+    if (hasIndex) {
+      index = new BAMFileIndexImpl(
+          IOUtil.maybeBufferedSeekableStream(indexStream),header.getSequenceDictionary());
+    } else {
+      index = null;
+    }
+  }
+  
+  void processHeader() {
+    contigsByReference = Maps.newHashMap();
+    for (Contig contig : requestedContigs) {
+      contigsByReference.put(contig.referenceName != null ? contig.referenceName : "", contig);
+    }
+    if (contigsByReference.size() == 0 || contigsByReference.containsKey("*")) {
+      LOG.info("Outputting unmapped reads shard ");
+      c.output(new BAMShard(filePath, null, new Contig("*", 0, -1)));
+    }
+    final boolean allReferences =
+        contigsByReference.size() == 0 || contigsByReference.containsKey("");
+    LOG.info("All references = " + allReferences);
+    LOG.info("BAM has index = " + reader.hasIndex());
+    LOG.info("BAM has browseable index = " + reader.indexing().hasBrowseableIndex());
+    LOG.info("Class for index = " + reader.indexing().getIndex().getClass().getName());  
+  }
+  
+  Contig desiredContigForReference(SAMSequenceRecord reference) {
+    Contig contig = contigsByReference.get(reference.getSequenceName());
+    if (contig == null) {
+      if (allReferences) {
+        contig = new Contig(reference.getSequenceName(), 0, -1);
+      } else {
+        LOG.fine("No contig requested for reference " + reference.getSequenceName());
+        return null;
+      }
+    }
+    assert contig != null;
+    if (contig.end <= 0) {
+      contig = new Contig(contig.referenceName, contig.start, reference.getSequenceLength());
+      LOG.info("Updated length for contig for " + contig.referenceName + " to " + contig.end);
+    }
+    return contig;
+  }
+  
+  void createShardsForReference(SAMSequenceRecord reference, Contig contig) {
+    final BitSet overlappingBins = GenomicIndexUtil.regionToBins(
+        (int) contig.start, (int) contig.end);
+    if (overlappingBins == null) {
+      LOG.warning("Now overlapping bins for " + contig.start + ":" + contig.end);
+      return;
+    }
+    
+    BAMShard currentShard = null;
+    for (int binIndex = overlappingBins.nextSetBit(0); binIndex >= 0; binIndex = overlappingBins.nextSetBit(binIndex + 1)) {
+      final Bin bin = index.getBinData(reference.getSequenceIndex(), binIndex);
+      if (bin == null) {
+        continue;
+      }
+      if (LOG.isLoggable(Level.FINE)) {
+        LOG.fine("Processing bin " + index.getFirstLocusInBin(bin) + "-"
+            + index.getLastLocusInBin(bin));
+      }
+      
+      if (index.getLevelForBin(bin) != (GenomicIndexUtil.LEVEL_STARTS.length - 1)) {
+        if (LOG.isLoggable(Level.FINEST)) {
+          LOG.finest("Skipping - not lowest");
+        }
+        continue; // Skip non-lowest level bins
+      }
```

Comment:
@droazen Yeas we can because later on in the BMShard generation we call getSpanOverlapping that will make sure we get short reads that overlap boundaries.